### PR TITLE
split txcache from wallet json

### PR DESCRIPTION
**Pull request by wolverineks**
load transaction separate from `WalletJSON`
store transactions individually

### CODE REVIEW

File: src/SelectedWallet/WalletSelection/WalletSelectionScreen.tsx [hunk 1]
```diff
@@ -59,7 +59,7 @@ export const WalletSelectionScreen = () => {
         closeWallet()
         logout()
       } else {
-        throw error
+        Alert.alert('QWEQWE', error.message)
```
Comment (banklesss) [inline src/SelectedWallet/WalletSelection/WalletSelectionScreen.tsx]: QWEQWE ??

Comment (wolverineks) [reply src/SelectedWallet/WalletSelection/WalletSelectionScreen.tsx]: just a placeholder

Comment (wolverineks) [inline src/SelectedWallet/WalletSelection/WalletSelectionScreen.tsx]: what should we do here?

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 2]
```diff
@@ -876,3 +857,22 @@ export class ShelleyWallet extends Wallet implements WalletInterface {
 }
 
 const toHex = (bytes: Uint8Array) => Buffer.from(bytes).toString('hex')
+
+const makeStorageWithPrefix = (prefix: string): Storage => {
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts]: this Storage type shouldn't be declared in here?

File: src/yoroi-wallets/cardano/shelley/transactionCache.ts [hunk 3]
```diff
@@ -1,156 +1,72 @@
...
-  const addressToTxs = {}
+export class TransactionCache {
+  #state: TransactionCacheState
+  #subscriptions: Array<(transactions: Record<string, Transaction>) => void> = []
+  #perAddressTxsSelector = defaultMemoize(perAddressTxsSelector)
+  #perAddressCertificatesSelector = defaultMemoize(perAddressCertificatesSelector)
+  #confirmationCountsSelector = defaultMemoize(confirmationCountsSelector)
+  #storage: TxCacheStorage
+
+  static async create(storage: Pick<AsyncStorageStatic, 'getItem' | 'multiGet' | 'setItem' | 'multiSet'>) {
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/shelley/transactionCache.ts:33]: the storage type here can actually be any storage or it needs to be the storage with prefix?

File: src/SelectedWallet/WalletSelection/WalletSelectionScreen.tsx (hunk 1 above)
Comment (stackchain) [reply src/SelectedWallet/WalletSelection/WalletSelectionScreen.tsx]: Opening the wallet is no longer throwing authentication errors. For the invalid state or unrecoverable wallet, we need to get the text copy with product.

File: src/yoroi-wallets/cardano/shelley/transactionCache.test.ts [hunk 4]
```diff
@@ -1,166 +1,332 @@
...
+      expect(txCache.transactions).toMatchSnapshot()
+      expect(txCache.perAddressTxs).toMatchSnapshot()
+      expect(txCache.perRewardAddressCertificates).toMatchSnapshot()
+      expect(txCache.confirmationCounts).toMatchSnapshot()
 
-      expect(result).toBeUndefined()
-      expect(params.api.getTipStatus).toBeCalledTimes(1)
+      txCache.resetState()
+
+      expect(txCache.transactions).toMatchSnapshot()
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/shelley/transactionCache.test.ts:22]: resetting the state, should match the snapshot?

File: src/yoroi-wallets/cardano/shelley/transactionCache.test.ts [hunk 5]
```diff
@@ -1,166 +1,332 @@
...
+    const result = await syncTxs(params)
 
-      const result = await syncTxs(params)
+    expect(result).toBeUndefined()
+    expect(params.api.fetchNewTxHistory).toBeCalledTimes(3)
+  })
 
-      expect(result).toBeUndefined()
-    })
+  it('should return current txs plus new txs if there are new transactions', async () => {
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/shelley/transactionCache.test.ts:218]: shouldn't it test if the subscribers were called?

File: src/yoroi-wallets/cardano/shelley/transactionCache.ts [hunk 6]
```diff
@@ -1,156 +1,72 @@
...
-  return addressToPerTxCerts
-}
-
-const confirmationCountsSelector = (state: TransactionCacheState) => {
-  const {perAddressSyncMetadata, transactions} = state
-  return mapValues(transactions, (tx: Transaction) => {
-    if (tx.status !== TRANSACTION_STATUS.SUCCESSFUL) {
-      // TODO(ppershing): do failed transactions have assurance?
-      return null
+  private constructor({storage, transactions}: {storage: TxCacheStorage; transactions: Record<string, Transaction>}) {
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/shelley/transactionCache.ts:52]: Unfortunately, JS doesn't support private constructors https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields#:~:text=However%2C%20constructors%20cannot%20be%20private,to%20use%20a%20private%20flag.

Comment (stackchain) [reply src/yoroi-wallets/cardano/shelley/transactionCache.ts:52]: dunno if there is any impact if left here just as annotation

File: src/yoroi-wallets/cardano/shelley/transactionCache.ts [hunk 7]
```diff
@@ -505,3 +362,162 @@ export function toCachedTx(tx: RawTransaction): Transaction {
...
+
+const makeTxCacheStorage = (storage: Storage): TxCacheStorage => ({
+  loadTxs: async () => {
+    const txids: Array<string> = await storage.getItem('txids').then(parseTxids)
+    if (!txids) return {}
+    if (txids.length === 0) {
+      return {}
+    }
+
+    const values = await storage.multiGet(txids)
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/shelley/transactionCache.ts:478]: will this perform well without pagination?

**Review by stackchain — COMMENTED** [review]
lgfm, would like to test it with a big wallet to see how multiGet/multiSet

File: src/yoroi-wallets/cardano/shelley/transactionCache.test.ts (hunk 4 above)
Comment (wolverineks) [reply src/yoroi-wallets/cardano/shelley/transactionCache.test.ts:22]: the snapshots are indexed, each `matchSnapshot` is a different snapshot
it just means, "make sure whatever returns from this specific invocation never changes"

Comment (wolverineks) [reply src/yoroi-wallets/cardano/shelley/transactionCache.test.ts:22]: if you open up the branch in vscode, each `toMatchSnapshot` links to the corresponding snapshot

Comment (wolverineks) [reply src/yoroi-wallets/cardano/shelley/transactionCache.test.ts:22]: i use this method when refactoring, i dont know what it is supposed to be, but i dont want my changes to affect the return values

File: src/yoroi-wallets/cardano/shelley/transactionCache.test.ts (hunk 5 above)
Comment (wolverineks) [reply src/yoroi-wallets/cardano/shelley/transactionCache.test.ts:218]: this is testing `syncTxs`, not `TransactionCache`
this is testing implementation details, but i think we should change it in another pr
```
const txCache = TransactionCache.create(...)
const callback = jest.fn()
txCache.subscribe(callback)
await txCache.doSync(...)
expect(txCache.transaction)....
expect(callback).toHaveBeenCalledWith(...)
```

File: src/yoroi-wallets/cardano/shelley/transactionCache.ts (hunk 3 above)
Comment (wolverineks) [reply src/yoroi-wallets/cardano/shelley/transactionCache.ts:33]: any AsyncStorage

File: src/yoroi-wallets/cardano/shelley/transactionCache.ts (hunk 6 above)
Comment (wolverineks) [reply src/yoroi-wallets/cardano/shelley/transactionCache.ts:52]: there are private constructors in typescript
https://www.typescriptlang.org/play?#code/MYGwhgzhAEBiD29oG8BQ0PTALmgOwFcBbAIwFMAnadTABwoEsA3MAFzOmHjwlYoOCt4VABQ58xchQCUKGpkysAFgwgA6MNAC8WedAC+qPbzYNgnCmTZkxuQqUqy0ChXjIB3OIjHS9hw6hcPKzQAGaI2vgeXvAiAIwATADMvkG8YYhxkQjwasCW1vHJ0kA

File: src/yoroi-wallets/cardano/shelley/transactionCache.ts (hunk 7 above)
Comment (wolverineks) [reply src/yoroi-wallets/cardano/shelley/transactionCache.ts:478]: works fine for me on a simulator

File: storybook/mocks/wallet.ts [hunk 8]
```diff
@@ -75,7 +75,7 @@ export const mockWallet: YoroiWallet = {
     return Promise.resolve({isRegistered: false, poolKeyHash: null})
   },
   subscribeOnTxHistoryUpdate: () => {
-    null
+    return () => null
```
Comment (wolverineks) [inline storybook/mocks/wallet.ts:78]: returns unsubscribe now

File: src/yoroi-wallets/walletManager.ts [hunk 9]
```diff
@@ -311,7 +312,8 @@ export class WalletManager {
     const newWalletMeta = {...walletMeta}
 
     await wallet.restore(data, walletMeta)
-    wallet.id = walletMeta.id
+    if (!isYoroiWallet(wallet)) throw new Error('invalid wallet')
+
```
Comment (wolverineks) [inline src/yoroi-wallets/walletManager.ts:316]: moved assigning id into ShelleyWallet

File: src/yoroi-wallets/utils/versioning.ts [hunk 10]
```diff
@@ -8,13 +8,9 @@
...
-export const versionCompare = (versionA: string, versionB: string): number => {
-  for (const s of [versionA, versionB]) {
-    const re = /^\d+(.\d+){0,2}$/ // only accept numbers and dots
-    if (!(typeof s === 'string' || (s as unknown) instanceof String) || !s.match(re)) {
-      throw new Error(`versionCompare: invalid argument ${s}`)
-    }
-  }
+export type Version = `${number}.${number}.${number}` | `${number}.${number}` | `${number}`
+
+export const versionCompare = (versionA: Version, versionB: Version): number => {
```
Comment (wolverineks) [inline src/yoroi-wallets/utils/versioning.ts:13]: dont use throw for logic
typechecking and format checking with typescript as much as possible

File: src/yoroi-wallets/cardano/types.ts [hunk 11]
```diff
@@ -123,7 +123,7 @@ export interface WalletInterface {
   // =================== subscriptions =================== //
 
   subscribe(handler: (event: WalletEvent) => void): () => void
-  subscribeOnTxHistoryUpdate(handler: () => void): void
+  subscribeOnTxHistoryUpdate(handler: () => void): () => void
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/types.ts:126]: added unsubscribe

File: src/yoroi-wallets/cardano/shelley/transactionCache.ts [hunk 12]
```diff
@@ -162,98 +79,39 @@ export class TransactionCache {
...
+        bestBlockNum: this.#state.bestBlockNum,
         // @deprecated
-        perAddressSyncMetadata: this._state.perAddressSyncMetadata,
+        perAddressSyncMetadata: this.#state.perAddressSyncMetadata,
       })
-
-      Logger.info('TransactionCache: notifying subscribers on txs update')
-      this.notifyOnTxHistoryUpdate()
     }
   }
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/shelley/transactionCache.ts:114]: `toJSON`, `fromJSON`, cute but ultimately not useful

File: src/TxHistory/LockedDeposit.tsx [hunk 13]
```diff
@@ -73,30 +70,3 @@ const useStrings = () => {
     lockedDeposit: intl.formatMessage(globalMessages.lockedDeposit),
   }
 }
```
Comment (wolverineks) [inline src/TxHistory/LockedDeposit.tsx:72]: moved to hooks

File: src/SelectedWallet/WalletSelection/WalletSelectionScreen.tsx [hunk 14]
```diff
@@ -32,34 +30,31 @@ export const WalletSelectionScreen = () => {
...
 
-  useFocusEffect(closeWallet)
-
   const {openWallet, isLoading} = useOpenWallet({
-    onSuccess: ({wallet, walletMeta}) => {
+    onSuccess: ([wallet, walletMeta]) => {
       selectWalletMeta(walletMeta)
       selectWallet(wallet)
-      wallet.subscribeOnTxHistoryUpdate(() => queryClient.invalidateQueries([wallet.id, 'lockedAmount']))
+
```
Comment (wolverineks) [inline src/SelectedWallet/WalletSelection/WalletSelectionScreen.tsx:42]: moved to `useLockedAmount`

File: src/SelectedWallet/WalletSelection/WalletSelectionScreen.tsx [hunk 15]
```diff
@@ -32,34 +30,31 @@ export const WalletSelectionScreen = () => {
...
       } else if (error instanceof InvalidState) {
-        closeWallet()
         await showErrorDialog(errorMessages.walletStateInvalid, intl)
       } else if (error instanceof KeysAreInvalid) {
         await showErrorDialog(errorMessages.walletKeysInvalidated, intl)
-        closeWallet()
         logout()
       } else {
-        throw error
+        await showErrorDialog(errorMessages.generalError, intl, {message: error.message})
```
Comment (wolverineks) [inline src/SelectedWallet/WalletSelection/WalletSelectionScreen.tsx:57]: uncaught error changed to generic error dialog

File: src/SelectedWallet/WalletSelection/WalletSelectionScreen.tsx [hunk 16]
```diff
@@ -231,34 +226,6 @@ const OnlyDevButton = () => {
   return <Button onPress={() => navigation.navigate('developer')} title="Dev options" style={styles.button} />
 }
```
Comment (wolverineks) [inline src/SelectedWallet/WalletSelection/WalletSelectionScreen.tsx:228]: moved to hooks
`closeWallet` moved to `walletManager.openWallet`

File: src/yoroi-wallets/walletManager.ts [hunk 17]
```diff
@@ -300,6 +300,7 @@ export class WalletManager {
   }
 
   async openWallet(walletMeta: WalletMeta): Promise<[YoroiWallet, WalletMeta]> {
+    await this.closeWallet()
```
Comment (wolverineks) [inline src/yoroi-wallets/walletManager.ts:303]: currently, only 1 wallet open at a time
moved from `useOpenWallet`

File: src/SelectedWallet/WalletSelection/WalletSelectionScreen.tsx (hunk 1 above)
Comment (wolverineks) [reply src/SelectedWallet/WalletSelection/WalletSelectionScreen.tsx]: changed to generic error dialog

Comment (wolverineks) [reply src/SelectedWallet/WalletSelection/WalletSelectionScreen.tsx]: changed to generic error dialog

File: src/yoroi-wallets/cardano/ShelleyWallet.ts (hunk 2 above)
Comment (wolverineks) [reply src/yoroi-wallets/cardano/ShelleyWallet.ts]: moved to `storage.ts`

File: src/yoroi-wallets/cardano/shelley/transactionCache.ts (hunk 6 above)
Comment (stackchain) [reply src/yoroi-wallets/cardano/shelley/transactionCache.ts:52]: awesome

**Review by stackchain — APPROVED** [review]

**Review by banklesss — APPROVED** [review]

**Pull request merged**